xla/service/gpu/transforms/windowed_einsum_handler.cc

```diff
@@ -492,6 +511,125 @@ absl::Status PostProcessUnrolledLoop(HloInstruction* loop, int64_t stream_id) {
+    shape = ShapeUtil::PrependMajorDimension(1, shape);
+
+    for (auto& inst : partial_accumulations) {
+      CHECK(inst->user_count() == 1);
+      HloInstruction* reshaped_partial = while_body->AddInstruction(
+          HloInstruction::CreateReshape(shape, inst));
+      partials_to_concat.push_back(reshaped_partial);
+    }
+    Shape concat_shape = partial_accumulations[0]->shape();
+    concat_shape = ShapeUtil::PrependMajorDimension(
+        partial_accumulations.size(), concat_shape);
+
+    HloInstruction* concat = while_body->AddInstruction(
+        HloInstruction::CreateConcatenate(concat_shape, partials_to_concat, 0));
+
+    HloComputation* comp = loop->parent();
+    HloInstruction* windowed_lhs = loop->mutable_operand(0)->mutable_operand(0);
+    // Add a broadcasted zero of the same type as windowed_lhs. This holds all
+    // the partial accumulations and will be fed to a global reduction after
+    // this windowed einsum loop. We move the reduction outside of the loop so
+    // it can be fused or overlap with other instructions in the main
+    // computation.
+    Literal zero_literal =
+        LiteralUtil::Zero(windowed_lhs->shape().element_type());
+    HloInstruction* zero = comp->AddInstruction(
+        HloInstruction::CreateConstant(std::move(zero_literal)));
+    Shape zero_bcast_shape = ShapeUtil::ChangeElementType(
+        concat_shape, windowed_lhs->shape().element_type());
+    HloInstruction* zero_bcast = MakeBroadcastHlo(zero, {}, zero_bcast_shape);
+    loop->mutable_operand(0)->AppendOperand(zero_bcast);
+    ShapeUtil::AppendShapeToTuple(zero_bcast->shape(),
+                                  loop->mutable_operand(0)->mutable_shape());
+
+    // Update the parameter tuples of while's body and condition
+    // computations.
+    for (HloComputation* while_comp : {while_body, loop->while_condition()}) {
+      while_comp->ReplaceParameter(
+          0, HloInstruction::CreateParameter(
+                 0, loop->mutable_operand(0)->shape(),
+                 while_comp->parameter_instruction(0)->name()));
+    }
+    HloInstruction* root = while_body->root_instruction();
+    std::vector<HloInstruction*> original_operands(root->operands().begin(),
+                                                   root->operands().end());
+    original_operands.push_back(concat);
+    HloInstruction* new_output_tuple = while_body->AddInstruction(
+        HloInstruction::CreateTuple(original_operands));
+    TF_RETURN_IF_ERROR(while_body->ReplaceInstructionWithDifferentShape(
+        root, new_output_tuple));
+
+    // Update the shape of the while loop instruction.
+    *loop->mutable_shape() = loop->operand(0)->shape();
+
+    // The final reduction
+    HloInstruction* concat_result_gte =
+        comp->AddInstruction(HloInstruction::CreateGetTupleElement(
+            loop, (loop->operand(0)->shape().tuple_shapes_size() - 1)));
+    HloInstruction* reduced_result =
```

Comment:
It might be simpler to add the reduction in the while-loop, instead of outside of it. That way, you don't have to add a new tuple element to the input and output of the while-loop, but instead can replace the accumulated result (with tuple index 2) with the new reduced result.

Reply:
The reason for the reduction to be moved outside is to allow it to overlap with some other collectives or fused with other kernels. Due to data dependency, the reduction is always at the end of the loop and since the scheduler cannot schedule across computations, there's a good amount of compute space to hide some other collectives.